File: Functions_1/collatz.py
"""
The Collatz Sequence
Write a function named collatz() that has one parameter named number. If number is even, then collatz() should print number // 2 and return this value. If number is odd, then collatz() should print and return 3 * number + 1.

Then write a program that lets the user type in an integer and that keeps calling collatz() on that number until the function returns the value 1. (Amazingly enough, this sequence actually works for any integer—sooner or later, using this sequence, you’ll arrive at 1! Even mathematicians aren’t sure why. Your program is exploring what’s called the Collatz sequence, sometimes called “the simplest impossible math problem.”)

Remember to convert the return value from input() to an integer with the int() function; otherwise, it will be a string value.

Hint: An integer number is even if number % 2 == 0, and it’s odd if number % 2 == 1.

The output of this program could look something like this:

Enter number:
3
10
5
16
8
4
2
1


Input Validation
Add try and except statements to the previous project to detect whether the user types in a noninteger string. Normally, 
the int() function will raise a ValueError error if it is passed a noninteger string, as in int('puppy'). 
In the except clause, print a message to the user saying they must enter an integer.
"""

import time


def collatz(number):
    if number % 2 == 0:
        result = number // 2
    else:
        result = 3 * number + 1
    print(result)
    return result


def timed_collatz_sequence():
    while True:
        try:
            number = int(input("Enter number: "))
            start_time = time.time()

            while number != 1:
                number = collatz(number)

            end_time = time.time()
            elapsed_time = end_time - start_time

            print(f"Time taken: {elapsed_time:.6f} seconds")
            break  # Exit the loop if successful

        except ValueError:
            print("Error: You must enter an integer.")
            print("Please try again.")


# Run the timed callatz function
timed_collatz_sequence()

File: Functions/collatz.rs
use std::io;
use std::time::Instant;

// The Collatz function: If n is even, divide by 2. If n is odd, multiply by 3 and add 1.
fn collatz(number: i64) -> i64 {
    if number % 2 == 0 {
        number / 2
    } else {
        3 * number + 1
    }
}

fn timed_collatz_sequence() {
    loop {
        println!("Enter number: ");
        let mut input = String::new();

        // Read user input from stdin
        io::stdin()
            .read_line(&mut input)
            .expect("Failed to read line");

        // Parse the input and handle potential errors
        match input.trim().parse() {
            Ok(mut number) => {
                // Start timing the sequence
                let start_time = Instant::now();

                // Continue applying the Collatz function until we reach 1
                while number != 1 {
                    number = collatz(number);
                    println!("{}", number);
                }

                // Calculate and print the elapsed time
                let elapsed_time = start_time.elapsed();
                println!("Time taken: {:.6} seconds", elapsed_time.as_secs_f64());
                break;
            }
            Err(_) => {
                // Handle invalid input
                println!("Error: You must enter an integer.");
                println!("Please try again.");
            }
        }
    }
}

fn main() {
    // Call the function to start the Collatz sequence calculation
    timed_collatz_sequence();
}
